ElDorito/Source/Patches/Ui.cpp

```diff
@@ -96,79 +221,440 @@ namespace Patches::Ui
+				break;
+			}
+
+		//If the speaking_player string cannot be found, RIP. This shouldn't happen unless tags don't have correct modifications.
+		if (speakingPlayerOffset != NULL)
+			speakingPlayerStringFound = true;
+		else
+		{
+			speakingPlayerStringFound = false;
+			UpdateVoiceChatHUD(false);
+		}
+	}
+
+	uint16_t teamBroadcastIndicatorIndex = 0; //Hud Widget containing Icons.
+	uint16_t broadcastAvailableIndex = 0; //Can Talk Icon
+	uint16_t broadcastIndex = 0; //Talking Icon.
+	uint16_t broadcastPTTIndex = 0; //Push To Talk Icon
+	uint16_t broadcastNoIndex = 0; //Can't Talk Icon
+	uint16_t spartanMotionTrackerIndex = 0; //motion_tracker hud widget.
+	uint16_t spartanVoiceBroadcast1Index = 0; //Speaking radar icon 1.
+	uint16_t spartanVoiceBroadcast2Index = 0; //Speaking radar icon 2.											
+	//Support for the elite HUD, just in case any mods need it.
+	uint16_t eliteMotionTrackerIndex = 0;
+	uint16_t eliteVoiceBroadcast1Index = 0; //Speaking radar icon 1.
+	uint16_t eliteVoiceBroadcast2Index = 0; //Speaking radar icon 2.
+	bool voiceChatIconValidTags = false;
+	void FindVoiceChatIconsTagData()
+	{
+		using Blam::Tags::TagInstance;
+		using Blam::Tags::UI::ChudDefinition;
+
+		if (!TagInstance::IsLoaded('chdt', scoreboardChdtIndex))
+			return;
+		else if (!TagInstance::IsLoaded('chdt', spartanChdtIndex))
+			return;
+		/*else if(!TagInstance::IsLoaded('chdt' eliteChdtIndex))
+			return*/
+
+		auto *scoreboardChud = Blam::Tags::TagInstance(scoreboardChdtIndex).GetDefinition<Blam::Tags::UI::ChudDefinition>();
+		auto *spartanChud = Blam::Tags::TagInstance(spartanChdtIndex).GetDefinition<Blam::Tags::UI::ChudDefinition>();
+		//auto *eliteChud = Blam::Tags::TagInstance(eliteChdtIndex).GetDefinition<Blam::Tags::UI::ChudDefinition>();
+
+		//If it's the first time updating the HUD, find the tagblock indexes.
+		//Find widgets in spartan.
+		for (int hudWidgetBlock = 0; hudWidgetBlock < spartanChud->HudWidgets.Count; hudWidgetBlock++)
+		{
+			if (spartanChud->HudWidgets[hudWidgetBlock].NameStringID == 0x2A76) //motion_tracker
+			{
+				spartanMotionTrackerIndex = hudWidgetBlock;
+
+				for (int bitmapWidgetBlock = 0; bitmapWidgetBlock < spartanChud->HudWidgets[hudWidgetBlock].BitmapWidgets.Count; bitmapWidgetBlock++)
+				{
+					if (spartanChud->HudWidgets[hudWidgetBlock].BitmapWidgets[bitmapWidgetBlock].NameStringID == 0x2A77) //voice_broadcast1
+						spartanVoiceBroadcast1Index = bitmapWidgetBlock;
+					else if (spartanChud->HudWidgets[hudWidgetBlock].BitmapWidgets[bitmapWidgetBlock].NameStringID == 0x2A78) //voice_broadcast2
+						spartanVoiceBroadcast2Index = bitmapWidgetBlock;
+
+					if (spartanVoiceBroadcast1Index != NULL && spartanVoiceBroadcast2Index != NULL)
+						break;
+				}
+			}
+			if (spartanMotionTrackerIndex != NULL)
+				break;
+		}
+
+		/*
+		//Find widgets in elite.
+		for (int hudWidgetBlock = 0; hudWidgetBlock < eliteChud->HudWidgets.Count; hudWidgetBlock++)

```

Comment:
When this is implemented in the future we can add Chud array to remove duplicate code